# TensorFlow Assignment: Multilayer Perceptron (MLP) and Convolutional Neural Network (CNN)

**[Duke Community Standard](http://integrity.duke.edu/standard.html): By typing your name below, you are certifying that you have adhered to the Duke Community Standard in completing this assignment.**

Name: [Bohao Huang]

Now that you've run through a simple logistic regression model on MNIST, let's see if we can do better (Hint: we can). For this assignment, you'll build a multilayer perceptron (MLP) and a convolutional neural network (CNN), two popular types of neural networks, and compare their performance. Some potentially useful code:

In [1]:
import tensorflow as tf
from tensorflow.examples.tutorials.mnist import input_data

# Import data
mnist = input_data.read_data_sets("MNIST_data/", one_hot=True)

Extracting MNIST_data/train-images-idx3-ubyte.gz
Extracting MNIST_data/train-labels-idx1-ubyte.gz
Extracting MNIST_data/t10k-images-idx3-ubyte.gz
Extracting MNIST_data/t10k-labels-idx1-ubyte.gz


In [2]:
# Helper functions for creating weight variables
def weight_variable(shape):
    """weight_variable generates a weight variable of a given shape."""
    initial = tf.truncated_normal(shape, stddev=0.1)
    return tf.Variable(initial)

def bias_variable(shape):
    """bias_variable generates a bias variable of a given shape."""
    initial = tf.constant(0.1, shape=shape)
    return tf.Variable(initial)

# Tensorflow Functions that might also be of interest:
# tf.nn.sigmoid()
# tf.nn.relu()

### Multilayer Perceptron

Build a multilayer perceptron for MNIST digit classfication. Feel free to play around with the model architecture and see how the training time/performance changes, but to begin, try the following:

Image -> fully connected (500 hidden units) -> nonlinearity (Sigmoid/ReLU) -> fully connected (10 hidden units) -> softmax

Skeleton framework for you to fill in (Code you need to provide is marked by `###`):

### Sigmoid Version

In [3]:
# Model Inputs
x = tf.placeholder(tf.float32, [None, 784])
y_ = tf.placeholder(tf.float32, [None, 10])

# Define the graph
### Create your MLP here##
### Make sure to name your MLP output as y_mlp ###
fc1_w = weight_variable([784, 500])
fc1_b = bias_variable([500])
fc2_w = weight_variable([500, 10])
fc2_b = bias_variable([10])

layer_fc1 = tf.matmul(x, fc1_w) + fc1_b
layer_nonlinear = tf.sigmoid(layer_fc1)
y_mlp = tf.matmul(layer_nonlinear, fc2_w) + fc2_b
# y_mlp = tf.nn.softmax(layer_fc2)


# Loss 
cross_entropy = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits(labels=y_, logits=y_mlp))

# Optimizer
train_step = tf.train.GradientDescentOptimizer(0.5).minimize(cross_entropy)

# Evaluation
correct_prediction = tf.equal(tf.argmax(y_mlp, 1), tf.argmax(y_, 1))
accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float32))

with tf.Session() as sess:
    # Initialize all variables
    sess.run(tf.global_variables_initializer())
    
    # Training regimen
    for i in range(4000):
        # Validate every 250th batch
        if i % 250 == 0:
            validation_accuracy = 0
            for v in range(10):
                batch = mnist.validation.next_batch(100)
                validation_accuracy += (1/10) * accuracy.eval(feed_dict={x: batch[0], y_: batch[1]})
            print('step %d, validation accuracy %g' % (i, validation_accuracy))
        
        # Train    
        batch = mnist.train.next_batch(50)
        train_step.run(feed_dict={x: batch[0], y_: batch[1]})

    print('test accuracy %g' % accuracy.eval(feed_dict={x: mnist.test.images, y_: mnist.test.labels}))

step 0, validation accuracy 0.092
step 250, validation accuracy 0.9
step 500, validation accuracy 0.922
step 750, validation accuracy 0.926
step 1000, validation accuracy 0.934
step 1250, validation accuracy 0.947
step 1500, validation accuracy 0.927
step 1750, validation accuracy 0.952
step 2000, validation accuracy 0.949
step 2250, validation accuracy 0.953
step 2500, validation accuracy 0.96
step 2750, validation accuracy 0.951
step 3000, validation accuracy 0.964
step 3250, validation accuracy 0.962
step 3500, validation accuracy 0.963
step 3750, validation accuracy 0.964
test accuracy 0.9601


### ReLU Version

In [4]:
# Model Inputs
x = tf.placeholder(tf.float32, [None, 784])
y_ = tf.placeholder(tf.float32, [None, 10])

# Define the graph
### Create your MLP here##
### Make sure to name your MLP output as y_mlp ###
fc1_w = weight_variable([784, 500])
fc1_b = bias_variable([500])
fc2_w = weight_variable([500, 10])
fc2_b = bias_variable([10])

layer_fc1 = tf.matmul(x, fc1_w) + fc1_b
layer_nonlinear = tf.nn.relu(layer_fc1)
y_mlp = tf.matmul(layer_nonlinear, fc2_w) + fc2_b
# y_mlp = tf.nn.softmax(layer_fc2)


# Loss 
cross_entropy = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits(labels=y_, logits=y_mlp))

# Optimizer
train_step = tf.train.GradientDescentOptimizer(0.5).minimize(cross_entropy)

# Evaluation
correct_prediction = tf.equal(tf.argmax(y_mlp, 1), tf.argmax(y_, 1))
accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float32))

with tf.Session() as sess:
    # Initialize all variables
    sess.run(tf.global_variables_initializer())
    
    # Training regimen
    for i in range(4000):
        # Validate every 250th batch
        if i % 250 == 0:
            validation_accuracy = 0
            for v in range(10):
                batch = mnist.validation.next_batch(100)
                validation_accuracy += (1/10) * accuracy.eval(feed_dict={x: batch[0], y_: batch[1]})
            print('step %d, validation accuracy %g' % (i, validation_accuracy))
        
        # Train    
        batch = mnist.train.next_batch(50)
        train_step.run(feed_dict={x: batch[0], y_: batch[1]})

    print('test accuracy %g' % accuracy.eval(feed_dict={x: mnist.test.images, y_: mnist.test.labels}))

step 0, validation accuracy 0.105
step 250, validation accuracy 0.942
step 500, validation accuracy 0.953
step 750, validation accuracy 0.957
step 1000, validation accuracy 0.969
step 1250, validation accuracy 0.972
step 1500, validation accuracy 0.974
step 1750, validation accuracy 0.967
step 2000, validation accuracy 0.973
step 2250, validation accuracy 0.973
step 2500, validation accuracy 0.975
step 2750, validation accuracy 0.978
step 3000, validation accuracy 0.979
step 3250, validation accuracy 0.981
step 3500, validation accuracy 0.986
step 3750, validation accuracy 0.974
test accuracy 0.9799


#### Comparison

How do the sigmoid and rectified linear unit (ReLU) compare?

***

The ReLU version performs better than sigmoid version in these two ways:
- ReLU version converges faster: 0.934 accuracy at step 1000 while it is 0.969 at step 1000 for sigmoid version
- ReLU version have higher test accuracy (0.9799) than sigmoid version (0.9601)

***

### Convolutional Neural Network

Build a simple 2-layer CNN for MNIST digit classfication. Feel free to play around with the model architecture and see how the training time/performance changes, but to begin, try the following:

Image -> CNN (32 5x5 filters) -> nonlinearity (ReLU) ->  (2x2 max pool) -> CNN (64 5x5 filters) -> nonlinearity (ReLU) -> (2x2 max pool) -> fully connected (1024 hidden units) -> nonlinearity (ReLU) -> fully connected (10 hidden units) -> softmax

Some additional functions that you might find helpful:

In [5]:
# Convolutional neural network functions
def conv2d(x, W):
    """conv2d returns a 2d convolution layer with full stride."""
    return tf.nn.conv2d(x, W, strides=[1, 1, 1, 1], padding='SAME')

def max_pool_2x2(x):
    """max_pool_2x2 downsamples a feature map by 2X."""
    return tf.nn.max_pool(x, ksize=[1, 2, 2, 1], strides=[1, 2, 2, 1], padding='SAME')

# Tensorflow Function that might also be of interest:
# tf.reshape()

Skeleton framework for you to fill in (Code you need to provide is marked by `###`):

*Hint: Convolutional Neural Networks are spatial models. You'll want to transform the flattened MNIST vectors into images for the CNN. Similarly, you might want to flatten it again sometime before you do a softmax. You also might want to look into the  [conv2d() documentation](https://www.tensorflow.org/api_docs/python/tf/nn/conv2d) to see what shape kernel/filter it's expecting.*

In [6]:
# Model Inputs
x = tf.placeholder(tf.float32, [None, 784])
y_ = tf.placeholder(tf.float32, [None, 10])

# Define the graph
### Create your CNN here##
### Make sure to name your CNN output as y_conv ###

# first cnn
cnn1_w = weight_variable([5, 5, 1, 32])
cnn1_b = bias_variable([32])
# second cnn
cnn2_w = weight_variable([5, 5, 32, 64])
cnn2_b = bias_variable([64])
# first fc
fc1_w = weight_variable([64*7*7, 1024])
fc1_b = bias_variable([1024])
# second fc
fc2_w = weight_variable([1024, 10])
fc2_b = bias_variable([10])

# CNN (32 5x5 filters) -> nonlinearity (ReLU) -> (2x2 max pool)
layer_cnn1 = conv2d(tf.reshape(x, [-1, 28, 28, 1]), cnn1_w) + cnn1_b # must do a reshape here # 32*28*28
layer_relu1 = tf.nn.relu(layer_cnn1)
layer_maxpool1 = max_pool_2x2(layer_relu1) # 32*14*14
# CNN (64 5x5 filters) -> nonlinearity (ReLU) -> (2x2 max pool)
layer_cnn2 = conv2d(layer_maxpool1, cnn2_w) + cnn2_b # 64*14*14
layer_relu2 = tf.nn.relu(layer_cnn2)
layer_maxpool2 = max_pool_2x2(layer_relu2) # 64*7*7
# fully connected (1024 hidden units) -> nonlinearity (ReLU)
layer_fc1 = tf.matmul(tf.reshape(layer_maxpool2, [-1, 64*7*7]), fc1_w) + fc1_b
layer_relu3 = tf.nn.relu(layer_fc1)
# fully connected (10 hidden units) -> softmax
y_conv = tf.matmul(layer_relu3, fc2_w) + fc2_b
# y_conv = tf.nn.softmax(layer_fc2)


# Loss 
cross_entropy = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits(labels=y_, logits=y_conv))

# Optimizer
train_step = tf.train.AdamOptimizer(1e-4).minimize(cross_entropy)

# Evaluation
correct_prediction = tf.equal(tf.argmax(y_conv, 1), tf.argmax(y_, 1))
accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float32))

with tf.Session() as sess:
    # Initialize all variables
    sess.run(tf.global_variables_initializer())
    
    # Training regimen
    for i in range(10000):
        # Validate every 250th batch
        if i % 250 == 0:
            validation_accuracy = 0
            for v in range(10):
                batch = mnist.validation.next_batch(50)
                validation_accuracy += (1/10) * accuracy.eval(feed_dict={x: batch[0], y_: batch[1]})
            print('step %d, validation accuracy %g' % (i, validation_accuracy))
        
        # Train    
        batch = mnist.train.next_batch(50)
        train_step.run(feed_dict={x: batch[0], y_: batch[1]})

    print('test accuracy %g' % accuracy.eval(feed_dict={x: mnist.test.images, y_: mnist.test.labels}))

step 0, validation accuracy 0.064
step 250, validation accuracy 0.904
step 500, validation accuracy 0.95
step 750, validation accuracy 0.956
step 1000, validation accuracy 0.952
step 1250, validation accuracy 0.984
step 1500, validation accuracy 0.982
step 1750, validation accuracy 0.976
step 2000, validation accuracy 0.978
step 2250, validation accuracy 0.98
step 2500, validation accuracy 0.978
step 2750, validation accuracy 0.988
step 3000, validation accuracy 0.99
step 3250, validation accuracy 0.976
step 3500, validation accuracy 0.982
step 3750, validation accuracy 0.99
step 4000, validation accuracy 0.996
step 4250, validation accuracy 0.982
step 4500, validation accuracy 0.988
step 4750, validation accuracy 0.984
step 5000, validation accuracy 0.996
step 5250, validation accuracy 0.992
step 5500, validation accuracy 0.992
step 5750, validation accuracy 0.986
step 6000, validation accuracy 0.982
step 6250, validation accuracy 0.98
step 6500, validation accuracy 0.984
step 6750, v

Some differences from the logistic regression model to note:

- The CNN model might take a while to train. Depending on your machine, you might expect this to take up to half an hour. If you see your validation performance start to plateau, you can kill the training.

- The logistic regression model we used previously was pretty basic, and as such, we were able to get away with using the GradientDescentOptimizer, which performs implements the gradient descent algorithm. For more difficult optimization spaces (such as the ones deep networks pose), we might want to use more sophisticated algorithms. Prof David Carlson has a lecture on this later.
    
- Because of the larger size of our network, notice that our minibatch size has shrunk.
    
- We've added a validation step every 250 minibatches. This let's us see how our model is doing during the training process, rather than sit around twiddling our thumbs and hoping for the best when training finishes. This becomes especially significant as training regimens start approaching days and weeks in length. Normally, we validate on the entire validation set, but for the sake of time we'll just stick to 10 validation minibatches (500 images) for this homework assignment.

#### Comparison

How do the MLP and CNN compare in accuracy? Training time? Why would you use one vs the other? Is there a problem you see with MLPs when applied to other image datasets?

***

1. _How do the MLP and CNN compare in accuracy?_  
The CNN performs better than MLP considering of accuracy, it reaches plateu at around step 3750 and have an accuracy around 0.99. It has a test accuracy of 0.9877
2. _How do the MLP and CNN compare in training time?_  
The CNN consumes a much longer time than MLP. In fact it took my mac almost 45min to run the whole 10000 steps for CNN
3. _Why would you use one vs the other?_  
For a not very complex dataset like MNIST, I might prefer MLP for in consideration of training time. But for complex dataset/task I would choose CNN, though time consuming in training step, it is a much powerful model.

***